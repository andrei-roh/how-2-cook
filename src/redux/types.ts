export const SET_USER = 'SET_USER';
export const SET_LOADING = 'SET_LOADING';
export const SET_ERROR = 'SET_ERROR';
export const ADD_RECIPES_TO_LIST = 'ADD_RECIPES_TO_LIST';
export const UPDATE_RECIPES_LIST = 'UPDATE_RECIPES_LIST';
export const ADD_IMAGES_TO_LIST = 'ADD_IMAGES_TO_LIST';
export const SET_RECIPE_SEARCH_INPUT = 'SET_RECIPE_SEARCH_INPUT';
export const SET_MANAGE_RECIPES_SCROLL_SIZE = 'SET_MANAGE_RECIPES_SCROLL_SIZE';
export const SET_MANAGE_RECIPES_SCROLL_DIRECTION =
  'SET_MANAGE_RECIPES_SCROLL_DIRECTION';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';
export const ADD_INGREDIENTS_TO_LIST = 'ADD_INGREDIENTS_TO_LIST';
export const SET_INGREDIENT_SEARCH_INPUT = 'SET_INGREDIENT_SEARCH_INPUT';
export const UPDATE_INGREDIENTS_LIST = 'UPDATE_INGREDIENTS_LIST';


