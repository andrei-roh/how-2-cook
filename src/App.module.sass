@use 'src/style/colors'

.app 
  padding-top: 46px
  background-color: colors.$white
  height: 100vh
  box-sizing: border-box
  display: flex
  flex-direction: column
  align-items: center

