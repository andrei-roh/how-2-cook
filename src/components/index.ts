export { Header } from './Header/Header';
export { Button } from './Button/Button';
export { Input } from './Input/Input';
export { Modal } from './Modal/Modal';
export { Notification } from './Notification/Notification';
export { ImageLoader } from './ImageLoader/ImageLoader';
export { TextArea } from './TextArea/TextArea';
export { Select } from './Select/Select';