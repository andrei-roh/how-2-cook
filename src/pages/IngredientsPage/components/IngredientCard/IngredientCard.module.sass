@use 'src/style/colors'
@import 'src/style/typography'

.ingredientContainer
    width: 100%
    box-sizing: border-box
    padding: 16px
    border: solid 1px colors.$purple
    border-radius: 8px
    display: flex
    flex-direction: row
    justify-content: space-between
    gap: 10px

    .buttonsPanel
        display: flex
        gap: 15px

        .ingredientBlockLogo
            width: 24px