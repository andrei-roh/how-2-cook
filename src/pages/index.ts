export { TitlePage } from './TitlePage/TitlePage';
export { RecipesPage } from './RecipesPage/RecipesPage';
export { CreateRecipePage } from './CreateRecipePage/CreateRecipePage';
export { ShowRecipePage } from './ShowRecipePage/ShowRecipePage';
export { EditRecipePage } from './EditRecipePage/EditRecipePage';
export { IngredientsPage } from './IngredientsPage/IngredientsPage';
export { WeekPlanningPage } from './WeekPlanningPage/WeekPlanningPage';
export { MyPlansPage } from './MyPlansPage/MyPlansPage';