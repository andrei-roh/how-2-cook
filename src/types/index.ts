import { Action } from '@reduxjs/toolkit';

export interface IUser {
  email: string;
}

export enum DishType {
  Main = 'main',
  Soup = 'soup',
  Dessert = 'dessert',
  Salat = 'salat',
  Snack = 'snack',
}

export type RecipeImage = {
  [key: string]: string;
};

export interface IIngredient {
  id: string;
  name: string;
}

export interface IRecipe {
  id: string;
  imageUrl: string;
  name: string;
  type: DishType;
  isVegan: boolean;
  ingredients: string[];
  description: string;
  composition?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface IPlanningList {
  id: string;
  name: string;
  ingredientsList: string[];
  recipesList: string[];
}

export interface IState {
  user: IUser;
  loading: boolean;
  error: null | string;
  recipesList: IRecipe[];
  imagesList: RecipeImage[];
  recipesPage: {
    recipeSearchInput: string;
    scrollSize: number;
    scrollDirection: ScrollDirection;
  };
  isMenu: boolean;
  ingredientsList: IIngredient[];
  ingredientsPage: {
    ingredientSearchInput: string;
  }
  previousRoute: string;
  planningPage: {
    ingredientsList: string[];
    recipesList: IRecipe[];
  }
}

export interface IAction extends Action {
  type: string;
  payload: never;
}

export interface IValidationError {
  [key: string]: string;
}

export enum FirebaseErrorType {
  InvalidEmail = 'auth/invalid-email',
  InvalidPassword = 'auth/invalid-credential',
  TooManyRequsts = 'auth/too-many-requests',
}

export enum ImageType {
  Webp = 'image/webp',
  Jpeg = 'image/jpeg',
  Jpg = 'image/jpg',
  Png = 'image/png',
}

export interface IOption {
  name: string;
  value: string;
}

export enum ScrollDirection {
  Up = 'up',
  Down = 'down',
}

export enum Severity {
  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Error = 'error',
}

export interface ICurrentRecipe extends Partial<IRecipe> {}
