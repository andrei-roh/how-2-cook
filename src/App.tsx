import { Outlet } from 'react-router-dom';
import css from './App.module.sass';
import { Header, Notification } from './components';

const App = () => {
  return (
    <>
      <Header />
      <div className={css.app}>
        <Outlet />
      </div>
      <Notification />
    </>
  );
};

export default App;
