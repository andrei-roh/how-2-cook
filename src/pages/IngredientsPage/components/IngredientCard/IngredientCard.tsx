import css from './IngredientCard.module.sass';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';
import Edit from 'src/assets/pencil.svg';
import Bin from 'src/assets/paper-bin.svg';
import { Modal } from 'src/components';
import { IIngredient, IState } from 'src/types';
import { useState } from 'react';
import { deleteIngredient } from 'src/utils/deleteIngredient';
import { updateIngredientsList } from 'src/redux/actions';
import { useDispatch, useSelector } from 'react-redux';

interface IngredientCardProps {
  ingredient: IIngredient;
}

export const IngredientCard = ({ ingredient }: IngredientCardProps) => {
  const dispatch = useDispatch();
  const ingredientsList = useSelector((state: IState) => state.ingredientsList);
  const [isDelete, setIsDelete] = useState(false);

  const handleShowModal = () => setIsDelete(true);
  const handleCloseModal = () => setIsDelete(false);
  const handleDeleteIngredient = (id: string, name: string) => {
    deleteIngredient(id, name).then((result) => {
      if (result) {
        const deletedIngedientIndex = ingredientsList.findIndex(
          (ingredient) => ingredient.id === id
        );
        const updatedIngredientsList: IIngredient[] = [...ingredientsList];
        updatedIngredientsList.splice(deletedIngedientIndex, 1);

        dispatch(updateIngredientsList([...updatedIngredientsList]));
        handleCloseModal();
      }
    });
  };

  return (
    <>
      <Box
        key={ingredient.id}
        {...ingredient}
        className={css.ingredientContainer}
      >
        <Typography variant='button'>{ingredient.name}</Typography>
        <Box className={css.buttonsPanel}>
          <img
            onClick={() => null}
            className={css.ingredientBlockLogo}
            src={Edit}
            alt='Edit Ingredient Button'
          />
          <img
            onClick={handleShowModal}
            className={css.ingredientBlockLogo}
            src={Bin}
            alt='Delete Ingredient Button'
          />
        </Box>
      </Box>
      {isDelete && (
        <Modal
          cancelButtonMessage='Отмена'
          submitButtonMessage='Удалить'
          isLoading={false}
          handleClose={handleCloseModal}
          handleSubmit={() =>
            handleDeleteIngredient(ingredient.id, ingredient.name)
          }
          message={`Вы уверены, что хотите удалить ингредиент ${ingredient.name}?`}
        />
      )}
    </>
  );
};
