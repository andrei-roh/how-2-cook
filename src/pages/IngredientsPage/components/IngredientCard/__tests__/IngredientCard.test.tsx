import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { IIngredient, IUser } from 'src/types';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { IngredientCard } from '../IngredientCard';

const navigateMock = vi.fn();
const dispatchMock = vi.fn();
const userMock = { email: 'Test@email.com' };
const recipesMock: IIngredient = {
  id: '00000000-0000-0000-0000-000000000001',
  name: 'Test Ingredient 1',
};

const stateMock = {
  user: userMock as IUser,
  loading: false,
  error: null,
};

const reducerMock = (state = stateMock) => {
  return state;
};

const storeMock = configureStore({ reducer: reducerMock });

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => ({ navigate: navigateMock }),
  };
});

vi.mock('react-redux', async () => {
  const actual = await vi.importActual('react-redux');

  return {
    ...actual,
    useDispatch: () => ({ dispatch: dispatchMock }),
  };
});

describe('Testing: IngredientCard', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <IngredientCard ingredient={recipesMock} />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
