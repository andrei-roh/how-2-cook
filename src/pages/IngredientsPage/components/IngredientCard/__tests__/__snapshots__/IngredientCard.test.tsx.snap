// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: IngredientCard > should render 1`] = `
<div>
  <div
    class="_ingredientContainer_dd3edf MuiBox-root css-0"
    id="00000000-0000-0000-0000-000000000001"
    name="Test Ingredient 1"
  >
    <span
      class="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
    >
      Test Ingredient 1
    </span>
    <div
      class="_buttonsPanel_dd3edf MuiBox-root css-0"
    >
      <img
        alt="Edit Ingredient Button"
        class="_ingredientBlockLogo_dd3edf"
        src="/src/assets/pencil.svg"
      />
      <img
        alt="Delete Ingredient Button"
        class="_ingredientBlockLogo_dd3edf"
        src="/src/assets/paper-bin.svg"
      />
    </div>
  </div>
</div>
`;
